### ISSUE TITLE
"space-before-function-paren" should check for class constructors

### ISSUE THREAD

**Issue by ahmadmaged** [eslint/eslint#7688]
Hi,
I am using ESLint v3.10.2 without changing the parser.

The part of my configuration that is related to the problem is :
(this is NOT the complete .eslintrc.json file)
```js
{
    "env": {
        "browser": true,
        "node": true
    },

    "extends": "eslint:recommended",

    "rules": {
        "space-before-function-paren": [ "error", {
            "anonymous": "always",
            "named": "never",
            "asyncArrow": "ignore"
        } ]
    }
}
```

**Background :**
Our team requires a space between the keyword `function` and the following parentheses that contain the parameter list to maintain clarity and readability, such as :
```js
function () {
    /* code goes here */
}
```
The same style is maintained for keywords such as `class` & `constructor` (meaning a space between the keyword and the following parentheses or braces).

**The Problem :**
When defining a class constructor, ESLint reports an error if the keyword `constructor` is followed by a space.

**For Example :**
```js
/* eslint-env es6 */

class Food {

    constructor (name, protein, carbs, fat) {
        this.name = name;
        this.protein = protein;
        this.carbs = carbs;
        this.fat = fat;
    }

    toString() {
        return `${this.name}, ${this.protein}g Protein, ${this.carbs}g Carbohydrate, ${this.fat}g Fat`;
    }

    print() {
        console.log( this.toString() );
    }
}
```

Adding a space should be allowed to provide a consistent style for programmers who always include a space after keywords such as `class`, `function`, `constructor`.

Instead, I receive this error :
```sh
5:16  error    Unexpected space before function parentheses  space-before-function-paren
```
It seams like ESLint wrongly treats the keyword `constructor` as a method name and as a result reports an error if the method name is followed by a space (according to the configuration).

**Comment by not-an-aardvark** [comment]
Thanks for the report. However, I think this is working as intended.

> Adding a space should be allowed to provide a consistent style for programmers who always include a space after keywords such as class, function, constructor.

`constructor` is not a keyword. Syntactically, aside from a few edge cases, a `constructor` method is the same as any other method.

**Comment by ahmadmaged** [comment]
Well, I agree that "space-before-function-paren" is working as its description says.

But, is there a rule or a rule-option that would allow class constructors to be followed by a space ?

I am sure there are developers that like to add a space between a constructor and its parameter list.

If no rule or option exists for that, can it be added in a future version ?

Thanks for the rapid response.

**Comment by ahmadmaged** [comment]
Feel free to classify this as a feature request if necessary. :smile:

**Comment by not-an-aardvark** [comment]
Ok, I reclassified it as a feature request.

If we were going to allow exceptions with functions such as `constructor`, I think it would be better to allow the user to customize functions that should be exceptions, rather than just adding a boolean option for `constructor`. For example, we could support a regular expression option for function names. However, while I'm open to being convinced otherwise, I personally don't see this as a common enough use case to add an option to core.

If you're interested in seeing what the rest of the team things, would you mind filling out the [rule change proposal template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md)?

**Comment by ahmadmaged** [comment]
I agree that it would be a more useful idea to allow the user to customize some functions to follow a different behavior (to enforce or prevent spacing) through _1) a regular expression_ or _2) an array of strings_.

That would allow a broader group of use cases for the proposed option.

**Comment by kaicataldo** [comment]
I don't really understand why this is desirable. I've never seen this code style before and, as @not-an-aardvark mentioned, `constructor` is not a keyword. @ahmadmaged mind explaining?

**Comment by ahmadmaged** [comment]
@kaicataldo It seems more like a niche for me. :relieved:

Of course you are correct about `constructor` not being a keyword, just a special function.
Although, I must admit I wasn't aware of that until @not-an-aardvark gave me the heads-up.

I think the idea of allowing functions to follow a different spacing behavior by using a list or a regular expression might be more useful.

However, if the feature doesn't serve a common use-case for the broader community of eslint users, then I think it will be better implemented as a _plugin_ not as a part of _core eslint_.

**Comment by nzakas** [comment]
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

You can, of course, create your own rule and distribute it as a plugin if you still have a need.
